File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:

        res = []

        def backtrack(curr):

            #Base Case
            if len(curr) == len(nums):
                res.append(curr.copy())
                return

            #Recursive Case
            for i in range(len(nums)):

                if nums[i] in curr:
                    continue

                curr.append(nums[i])
                backtrack(curr)
                curr.pop()



        backtrack([])
        return res

File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        if (nums.empty())
        {
            return {};
        }
        
        int n = nums.size();
        vector<int> curr;
        vector<bool> used(n, false); 

        backtrack(nums, 0, curr, n, used);

        return res;
    }

    void backtrack(vector<int>& nums, int index, vector<int>& curr, int n, vector<bool>& used)
    {

        if (curr.size() == n)
        {
            res.push_back(curr);
        }

        if (index >= n)
        {
            return;
        }

        for (int i = 0; i < n; i++)
        {
            if (used[i])
            {
                continue;
            }

            used[i] = true;
            curr.push_back(nums[i]);
            backtrack(nums, i, curr, n, used);
            curr.pop_back();
            used[i] = false;
        }

    }

private:
    vector<vector<int>> res;
};